test_runner/regress/test_replication_lag.py

```diff
@@ -0,0 +1,56 @@
+import threading
+
+from fixtures.log_helper import log
+from fixtures.neon_fixtures import NeonEnv, PgBin, wait_replica_caughtup
+
+
+def test_replication_lag(neon_simple_env: NeonEnv, pg_bin: PgBin):
+    env = neon_simple_env
+    n_iterations = 60
+
+    # Use aggressive GC and checkpoint settings
+    tenant, _ = env.neon_cli.create_tenant(
+        conf={
+            "gc_period": "15 s",  # should not be smaller than wal_receiver_status_interval
+            "gc_horizon": f"{1024 ** 2}",
+            "checkpoint_distance": f"{1024 ** 2}",
+            "compaction_target_size": f"{1024 ** 2}",
+            # set PITR interval to be small, so we can do GC
+            "pitr_interval": "5 s",
+        }
+    )
+
+    def run_pgbench(connstr: str):
+        log.info(f"Start a pgbench workload on pg {connstr}")
+        pg_bin.run_capture(["pgbench", "-T60", connstr])
+
+    with env.endpoints.create_start(
+        branch_name="main", endpoint_id="primary", tenant_id=tenant
+    ) as primary:
+        pg_bin.run_capture(["pgbench", "-i", "-s10", primary.connstr()])
+
+        t = threading.Thread(target=run_pgbench, args=(primary.connstr(),), daemon=True)
+        t.start()
+
+        with env.endpoints.new_replica_start(origin=primary, endpoint_id="secondary") as secondary:
+            wait_replica_caughtup(primary, secondary)
+            for _ in range(1, n_iterations):
+                try:
+                    primary_lsn = primary.safe_psql_scalar(
+                        "SELECT pg_current_wal_flush_lsn()::text", log_query=False
+                    )
+                    secondary_lsn = secondary.safe_psql_scalar(
+                        "SELECT pg_last_wal_replay_lsn()", log_query=False
+                    )
+                    balance = secondary.safe_psql_scalar(
+                        "select sum(abalance) from pgbench_accounts"
+                    )
+                    log.info(
+                        f"primary_lsn={primary_lsn}, secondary_lsn={secondary_lsn}, balance={balance}"
+                    )
+                except Exception as error:
+                    print(f"Query failed: {error}")
+                    if "canceling statement due to conflict with recovery" not in str(error):
```

Comment:
Why conflict with recovery error is allowed if we enable hot standby feedback?

Reply:
Because without protocol changes we can request page at LSN far in the past (when page was last modified) and it can be beyond replica replay LSN and gc_cutoff.

Reply:
So I refreshed my knowledge about recovery conflict.
@hlinnaka and @MMeent please confirm: recovery conflict is still possible even if `hot_standby_feedback` is enabled. It is reported by `ResolveRecoveryConflictWithVirtualXID` when time of waiting for virtual xid exceeds `max_standby_streaming_delay` (default value is 30 sec).

If I increase it to 100 sec, then the problem is gone.
Another question his whether such delay (> 30 seconds) is normal?

Reply:
Right, recovery conflicts are certainly possible even with `hot_standby_feedback`. It only helps with VACUUM-related conflicts, and I don't think it's not bulletproof at that either. Dropping a database in the primary while a backend is running on it in the standby is one easy way to get a conflict.

Reply:
> Dropping a database in the primary while a backend is running on it in the standby is one easy way to get a conflict.

Sure, but we don't drop databases etc here, just run pgbench, that's why I'm wondering.

Reply:
> Sure, but we don't drop databases etc here, just run pgbench, that's why I'm wondering.

Please look at the following code in standby.c  `ResolveRecoveryConflictWithVirtualXID`:

```
			if (WaitExceedsMaxStandbyDelay(wait_event_info))
			{
				pid_t		pid;

				/*
				 * Now find out who to throw out of the balloon.
				 */
				Assert(VirtualTransactionIdIsValid(*waitlist));
				pid = CancelVirtualTransaction(*waitlist, reason);
```

It seems to be completely not related with `hot_standby_feedback`.
As far as I understand if hot-standby backend is waiting for some transaction executed by recovery more than `max_standby_streaming_delay` then we kill it.

Reply:
My only question is whether it is normal and expected that backed at hot-standby replica is blocked for more than 30 seconds while there is intensive update stream from primary....

Reply:
> Please look at the following code in standby.c ResolveRecoveryConflictWithVirtualXID:

Well, in my understanding hot_feedback should prevent the conflict in the first place by preventing vacuum on primary removing versions needed by standby. Unless we do some DDL etc which we don't here.

Reply:
Yes, `hot_standby_feedback` prevents vacuum at primary from removing versions needed by replica.
But this is completely different problems: replica is not lacking any version, but read-only query is canceled because it is waiting for wal receiver transaction for more than `max_standby_streaming_delay`.